spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java

```diff
@@ -233,8 +270,8 @@ public void removeMessageGroup(Object groupId) {
 					return;
 				}
 
-				groupUpperBound.release(groupIdToMessageGroup.get(groupId).size());
-				groupIdToMessageGroup.remove(groupId);
+				this.groupToUpperBound.remove(groupId).release(this.groupCapacity);
```

Comment:
I've revised this and it does not make sense for me to `release` that `UpperBound` any more.
If if remove the `group`, we should just clean everything around it.
If something is waiting for the permits (]addMessageToGroup]), it should create a new `UpperBound`.